csi/controller_server.go

```diff
@@ -325,10 +325,6 @@ func (cs *ControllerServer) publishVolume(volume *longhornclient.Volume, nodeID
 }
 
 func (cs *ControllerServer) publishSharedVolume(req *csi.ControllerPublishVolumeRequest, volume *longhornclient.Volume) (*csi.ControllerPublishVolumeResponse, error) {
-	if !volume.Ready || len(volume.Controllers) == 0 || volume.Controllers[0].Endpoint == "" {
-		return nil, status.Errorf(codes.Aborted, "The volume %s is not ready for workloads", req.GetVolumeId())
-	}
-
```

Comment:
Can you explain why we can skip the check `len(volume.Controllers) == 0 || volume.Controllers[0].Endpoint == ""`?

Reply:
Sure thing, I will explain it based on the examples below.
- volume detached, we can see that the engine endpoint is not available, since the engine is not running.
- volume attached in maintenance mode, we can see that the engine endpoint is not available, since the frontend is disabled.
- volume after expansion, we can see that the engine endpoint is not available and that the `frontend is still disabled`, since the expansion disables the frontend but there is nothing that re enables it after expansion, it's expected that the following attach call will re enable it, if we prevent that attach call we cannot re enable the frontend.
- volume attached, we can see that the engine endpoint is available

**Conclusion:**
An engines endpoint is only available while it's attached in non maintenance mode, using this condition outside of these circumstances is not valid to determine whether the volume is ready for workloads. 

You might be asking yourself why we need to check this at all, currently the csi driver has way to much logic and all of this should be handled by the volume-manager, see my comment in the dev channel in regard to making the driver dumber.

The goal should be to have the driver call attach on the api then poll & evaluate the result. But that requires structured error details, so we can act on them, see epic #1875 

Currently returning all these Aborted errors on transient states, just confuses the users since they think something is wrong. Instead it would be better to just call attach & poll since the poll would wait for the transient state to pass and return success as expected instead of kubernetes having to retry the whole csi call.


**volume detached**
```
controllers": [
{
"address": "",
"endpoint": "",
"hostId": "",
"instanceManagerName": "",
"name": "test-rwx-manual-e-cfd524c8",
"running": false,
},
],
"disableFrontend": false,
```

**volume attached in maintenance mode**
```
"controllers": [
{
"address": "10.42.1.251",
"endpoint": "",
"hostId": "jmoody-lh-dev-work1",
"instanceManagerName": "instance-manager-e-e16b4017",
"name": "test-rwx-maintenance-e-dfbc8076",
"running": true,
},
],
"disableFrontend": true,
```

**volume after expansion**
```
"controllers": [
{
"address": "",
"endpoint": "",
"hostId": "",
"instanceManagerName": "",
"name": "pvc-a9579ac8-0351-4c8a-8ba0-d24214fcd941-e-51fd4e2c",
"running": false,
},
],
"disableFrontend": true,
```

**volume attached**
```
"controllers": [
{
"address": "10.42.3.37",
"endpoint": "/dev/longhorn/pvc-a9579ac8-0351-4c8a-8ba0-d24214fcd941",
"hostId": "jmoody-lh-dev-work2",
"instanceManagerName": "instance-manager-e-b1eda47e",
"name": "pvc-a9579ac8-0351-4c8a-8ba0-d24214fcd941-e-51fd4e2c",
"running": true,
},
],
"disableFrontend": false,
```

Reply:
Thank you for the explanation